src/latency_aware_policy.cpp

```diff
@@ -0,0 +1,136 @@
+  limitations under the License.
+*/
+
+#include "latency_aware_policy.hpp"
+
+#include "get_time.hpp"
+#include "logger.hpp"
+
+namespace cass {
+
+void LatencyAwarePolicy::init(const SharedRefPtr<Host>& connected_host, const HostMap& hosts) {
+  copy_hosts(hosts, hosts_);
+  for (HostMap::const_iterator i = hosts.begin(),
+       end = hosts.end(); i != end; ++i) {
+    i->second->enable_latency_tracking(settings_.scale_ns, settings_.min_measured);
+  }
+  ChainedLoadBalancingPolicy::init(connected_host, hosts);
+}
+
+void LatencyAwarePolicy::register_handles(uv_loop_t* loop) {
+  calculate_min_average_task_ = PeriodicTask::start(loop,
+                                                    settings_.update_rate_ms,
+                                                    this,
+                                                    LatencyAwarePolicy::on_work,
+                                                    LatencyAwarePolicy::on_after_work);
+}
+
+void LatencyAwarePolicy::close_handles() {
+  if (calculate_min_average_task_ != NULL) {
+    PeriodicTask::stop(calculate_min_average_task_);
+  }
+}
+
+QueryPlan* LatencyAwarePolicy::new_query_plan(const std::string& connected_keyspace,
+                                              const Request* request,
+                                              const TokenMap& token_map) {
+  return new LatencyAwareQueryPlan(this,
+                                   child_policy_->new_query_plan(connected_keyspace, request, token_map));
+}
+
+void LatencyAwarePolicy::on_add(const SharedRefPtr<Host>& host) {
+  host->enable_latency_tracking(settings_.scale_ns, settings_.min_measured);
+  add_host(hosts_, host);
+  ChainedLoadBalancingPolicy::on_add(host);
+}
+
+void LatencyAwarePolicy::on_remove(const SharedRefPtr<Host>& host) {
+  remove_host(hosts_, host);
+  ChainedLoadBalancingPolicy::on_remove(host);
+}
+
+void LatencyAwarePolicy::on_up(const SharedRefPtr<Host>& host) {
+  add_host(hosts_, host);
+  ChainedLoadBalancingPolicy::on_up(host);
+}
+
+void LatencyAwarePolicy::on_down(const SharedRefPtr<Host>& host) {
+  remove_host(hosts_, host);
+  ChainedLoadBalancingPolicy::on_down(host);
+}
+
+SharedRefPtr<Host> LatencyAwarePolicy::LatencyAwareQueryPlan::compute_next() {
+  int64_t min = policy_->min_average_.load();
+  const Settings& settings = policy_->settings_;
+  int64_t now = uv_hrtime();
+
+  SharedRefPtr<Host> host;
+  while ((host = child_plan_->compute_next())) {
+    TimestampedAverage latency = host->get_current_average();
+
+    if (min < 0 ||
+        latency.average < 0 ||
+        latency.num_measured < settings.min_measured ||
+        (now - latency.timestamp) > settings.retry_period_ns) {
+      return host;
+    }
+
+    if (latency.average <= static_cast<int64_t>(settings.exclusion_threshold * min)) {
+      return host;
+    }
+
+    skipped_.push_back(host);
+  }
+
+  if (!skipped_.empty()) {
+    host = skipped_.back();
+    skipped_.pop_back();
```

Comment:
I think using LIFO in this phase ends up offering less favorable hosts (as planned by the child policy) first. Wondering if it would be worth just indexing the vector for FIFO order.